functionary/prompt_utils.py

```diff
@@ -0,0 +1,210 @@
+from typing import List, Dict, Any, Optional, Tuple
+import torch
+from enum import Enum
+
+
+class EndToken(str, Enum):
+    system = "<|END_OF_SYSTEM|>"
+    user = "<|END_OF_USER|>"
+    assistant = "<|END_OF_ASSISTANT|>"
+    function = "<|END_OF_FUNCTION_RESULT|>"
+    function_call = "<|END_OF_FUNCTION_CALL|>"
+
+
+def get_end_token_for_message(message: Dict) -> EndToken:
+    """this function is used for getting the end token for each message.
+    For example, if message["role"] == "user" --> return EndToken.user
+    if message["role"] == "assistant" and "function_call" in message --> EndTOken.function_call
+
+    Args:
+        message (Dict): A dictionary containing: role, content, function_call(optional)
+
+    Returns:
+        EndToken: End Token for this message, this will be appended to the end of the prompt for this message
+    """
+    role = message["role"]
+    if role == "user":
+        return EndToken.user
+    elif role == "system":
+        return EndToken.system
+    elif role == "function":
+        return EndToken.function
+    else:  # role = assistant
+        if "function_call" in message:
+            return EndToken.function_call
+        else:
+            return EndToken.assistant
+
+
+def get_text_from_message(message: Dict) -> str:
+    """convert a message to a string to be included in the prompt
+
+    Args:
+        message (Dict): A dictionary in OpenAI format (containing: role, content, function_call (optional))
+
+    Returns:
+        str: the string used in the final prompt of this message
+    """
+    stop_token = get_end_token_for_message(message).value
+    content = message.get("content", "")
+
+    if content is not None:
+        content = f"{content}{stop_token}"
+
+    if message["role"] == "system":
+        text = "system:\n{content}\n".format(content=content)
+
+    elif message["role"] == "function":
+        text = "function name={name}:\n{content}\n".format(name=message.get("name", ""), content=content)
+
+    elif message["role"] == "user" and content is None:
+        text = "user:\n"
+
+    elif message["role"] == "user":
+        text = "user:\n{content}\n".format(content=content)
+
+    elif message["role"] == "assistant":
+        function = None
+        arguments = None
+        if (
+            "function_call" in message
+        ):  # format of openai: {"role": assistant, "function_call": {"name": xxx, "arguments": xxx}}
+            assert content is None
+            function = message["function_call"]["name"]
+            arguments = message["function_call"]["arguments"] + stop_token
+            text = f"assistant to={function}:\n{arguments}\n"
+        elif content is not None:  # this is text content
+            text = f"assistant:\n{content}\n"
+        else:  # if no function call and content is None --> this is used at inference
+            text = "assistant"
+
+    # This chunk of code below is to assert that stop_token is added correctly
```

Comment:
This is basically tests. So instead of assertions, lets use seperate tests to make sure its working as we expected. We need a test for each role and for each combination.

Reply:
This is really important. Cause there are lots of edge-cases as you probably know

Reply:
Sure will check this in test